File: MATTHEW/leetcode/AIcoding/1.Gemini-2.5/rust_solutions/229-majority-element-ii.rs
impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> Vec<i32> {
        if nums.is_empty() {
            return Vec::new();
        }

        let mut cand1 = 0;
        let mut count1 = 0;
        let mut cand2 = 0;
        let mut count2 = 0;

        for &num in &nums {
            if num == cand1 {
                count1 += 1;
            } else if num == cand2 {
                count2 += 1;
            } else if count1 == 0 {
                cand1 = num;
                count1 = 1;
            } else if count2 == 0 {
                cand2 = num;
                count2 = 1;
            } else {
                count1 -= 1;
                count2 -= 1;
            }
        }

        count1 = 0;
        count2 = 0;
        for &num in &nums {
            if num == cand1 {
                count1 += 1;
            } else if num == cand2 {
                count2 += 1;
            }
        }

        let mut result = Vec::new();
        let n = nums.len();
        let threshold = n / 3;

        if count1 > threshold {
            result.push(cand1);
        }
        if count2 > threshold && cand1 != cand2 {
            result.push(cand2);
        }

        result
    }
}

File: leetcode/AIcoding/1.Gemini-2.5/python3_solutions/229-majority-element-ii.py
class Solution:
    def majorityElement(self, nums: list[int]) -> list[int]:
        if not nums:
            return []

        cand1, cand2 = 0, 1
        count1, count2 = 0, 0

        for num in nums:
            if num == cand1:
                count1 += 1
            elif num == cand2:
                count2 += 1
            elif count1 == 0:
                cand1 = num
                count1 = 1
            elif count2 == 0:
                cand2 = num
                count2 = 1
            else:
                count1 -= 1
                count2 -= 1

        result = []
        n = len(nums)

        count1 = 0
        count2 = 0
        for num in nums:
            if num == cand1:
                count1 += 1
            elif num == cand2:
                count2 += 1

        if count1 > n // 3:
            result.append(cand1)
        if count2 > n // 3 and cand1 != cand2:
            result.append(cand2)

        return result